### ISSUE TITLE
Potential new implementation of Interval is needed

### ISSUE THREAD

**Issue by Ducasse** [guillep/pharo-core#11044]
So people like Siren needs to access the stop value of an interval.

**Comment by johnhwoods** [comment]
As distinct from `Interval>>last` then?  Looking more closely, I'm not even sure there should even be an `stop` instance variable.  I notice that `(2 to: 9 by: 3) = (2 to: 10 by: 3)` but `(2 to: 9 by: 3) extent ~= (2 to: 10 by: 3) extent` as a result of this stop vs. last confusion.  Would it perhaps be more elegant to set `last` appropriately in `setFrom:to:by` when the `Interval` is created?

**Comment by Ducasse** [comment]
```
(1 to: 10 by: 2) last
>>> 9
```
so last is not stop and we should add some executable comments like this one in the method. 
I would not change the implementation of interval to avoid break it but more strengthening the tests and the API.

**Comment by johnhwoods** [comment]
Perhaps we should consider the possibility that the existing implementation of `Interval` means it is already broken:
```Smalltalk
| a b |
a := (2 to: 9 by: 3).
b := (2 to: 10 by: 3).
a = b & a hash = b hash
```

**Comment by johnhwoods** [comment]
I don't think there's any need to keep `stop` at all.  It is specified on creation and effectively creates a `last` when it does.  Beyond the private method, used on creation, to set start, stop and step, the following methods refer to `stop`:
```
Interval>>#storeOn:
Interval>>#hash
Interval>>#+
Interval>>#size
Interval>>#last
Interval>>#-
Interval>>#printOn:
Interval>>#asOpenInterval
Interval>>#rangeIncludes:
Interval>>#extent
```
All of these can be re-written to use `last` instead of `stop` - mostly without changes in functionality and, I would suggest, where the functionality does change, it changes from "broken" to fixed"

**Comment by johnhwoods** [comment]
`storeOn` -- should use `last` rather than `stop`
`printOn` -- ditto
`hash` -- needs to use `last` -- there's a good argument that current implementation is incorrect as it stands.
`+` -- makes no difference (but maybe `*` and `/` should also be implemented).
`-` -- ditto
`size` -- should use `last` instead of calculating it from `stop` -- and be re-written anyway, currently a crazy implementation
`last` -- just return new ivar `last` instead of calculating it from `stop`
`extent` -- just use new ivar `last` instead of calculating it from `stop`
`rangeIncludes` -- would return different but probably more useful results if it used `last` rather than `stop`
`openInterval` -- this is a mad method that just shouldn't exist as the following two examples show:
```Smalltalk
(1 to: 1 by: -1) asOpenInterval asArray >>> #(0 1 2)
(1 to: 1 by: 1) asOpenInterval asArray >>> #()
```

**Comment by jordanmontt** [comment]
You are probably right. Now this is another issue. 
Adding stop accessor is 
- simple
- does not break anything. 
So we should be conservative first.
Then in a second time we can have an alternate implementation of interval. 
We can have parametrized tests that verifies equivalence between the intervalNow and newInterval 
and we can introduce it in Pharo. 

So we renamed the issue and created a new one to introduce this accessor.

**Comment by johnhwoods** [comment]
Cool.  I'll leave the implementation of `stop` accessor for someone else and will propose some changes to Interval.
I want to start from the fundamental assertion that Interval>>size > 0 .  An interval must always have at least the start element.  Currently (2 to: 0) is empty and (2 to: 0 by: -1) isn't.  I think both of these should be #(2).  Who disagrees?

**Comment by jordanmontt** [comment]
I disagree. I think that ` (2 to: 0)` should be empty. Else, we will be doing magic and having unexpected results. If someone wants to have a decreasing interval is better to use `(2 to: 0 by: -1)`

**Comment by johnhwoods** [comment]
Don't buy it.  Empty intervals are a nonsense: they *are* the 'magic and unexpected results.'   If there is a need for an empty interval there should be just one, like the empty set.  It makes no sense to have an infinite number of non-equal empty intervals:
`(2 to: 1.9) = (1 to: 0) >>> false`

**Comment by johnhwoods** [comment]
Having an Empty Interval would also fix the `asOpenInterval` as it would just `^self`
